modules/basicfuncs/ip-funcs.c

```diff
@@ -40,3 +42,121 @@ tf_ipv4_to_int(LogMessage *msg, gint argc, GString *argv[], GString *result)
+  HostResolveOptions *host_resolve_options = (HostResolveOptions *)data;
+  host_resolve_options->use_dns = cfg_process_yesno(value);
+  return TRUE;
+}
+
+static gboolean
+_process_dns_cache(const gchar *option_name, const gchar *value, gpointer data, GError **error)
+{
+  HostResolveOptions *host_resolve_options = (HostResolveOptions *)data;
+  host_resolve_options->use_dns_cache = cfg_process_yesno(value);
+  return TRUE;
+}
+
+static gboolean
+_process_normalize_hostnames(const gchar *option_name, const gchar *value, gpointer data, GError **error)
+{
+  HostResolveOptions *host_resolve_options = (HostResolveOptions *)data;
+  host_resolve_options->normalize_hostnames = cfg_process_yesno(value);
+  return TRUE;
+}
+
+static gboolean
+_parse_host_resolve_options(gint *argc, gchar **argv[], HostResolveOptions *host_resolve_options, GError **error)
+{
+  gboolean result = FALSE;
+  GOptionEntry entries[] =
+  {
+    { "use-fqdn", 'f', 0, G_OPTION_ARG_CALLBACK, _process_use_fqdn, NULL, NULL },
+    { "use-dns", 'd', 0, G_OPTION_ARG_CALLBACK, _process_use_dns, NULL, NULL },
+    { "dns-cache", 'c', 0, G_OPTION_ARG_CALLBACK, _process_dns_cache, NULL, NULL },
+    { "normalize-hostnames", 'n', 0, G_OPTION_ARG_CALLBACK, _process_normalize_hostnames, NULL, NULL },
+    { NULL }
+  };
+
+  GOptionContext *ctx = g_option_context_new(*argv[0]);
+  GOptionGroup *group = g_option_group_new("host-resolve-options", NULL, NULL, host_resolve_options, NULL);
+  g_option_group_add_entries(group, entries);
+  g_option_context_set_main_group(ctx, group);
+
+  if (!g_option_context_parse(ctx, argc, argv, error))
+    goto exit;
+
+  result = TRUE;
+exit:
+  g_option_context_free(ctx);
+  return result;
+}
+
+static gboolean
+_setup_host_resolve_options(DnsResolveIpState *state, gint *argc, gchar **argv[], GError **error)
+{
+  host_resolve_options_global_defaults(&state->host_resolve_options);
+  if (!_parse_host_resolve_options(argc, argv, &state->host_resolve_options, error))
+    return FALSE;
+
+  return TRUE;
+}
+
+static gboolean
+tf_dns_resolve_ip_prepare(LogTemplateFunction *self, gpointer s, LogTemplate *parent, gint argc, gchar *argv[],
+                          GError **error)
+{
+  DnsResolveIpState *state = (DnsResolveIpState *)s;
+
+  if (!_setup_host_resolve_options(state, &argc, &argv, error))
+    return FALSE;
+
+  if (argc > 2)
+    {
+      g_set_error(error, LOG_TEMPLATE_ERROR, 0,
+                  "dns-resolve-ip: too many arguments provided. usage: $(dns-resolve-ip [OPTIONS] IP)");
+      return FALSE;
+    }
+
+  if (!tf_simple_func_prepare(self, s, parent, argc, argv, error))
+    return FALSE;
+
+  return TRUE;
+}
+
+static void
+tf_dns_resolve_ip_call(LogTemplateFunction *self, gpointer s, const LogTemplateInvokeArgs *args, GString *result)
+{
+  DnsResolveIpState *state = (DnsResolveIpState *)s;
+  const gchar *hostname;
+  gsize result_len;
+
+  gchar *ip = args->argv[0]->str;
+  GSockAddr *addr = g_sockaddr_inet_or_inet6_new(ip, 0);
```

Comment:
We might want to introduce "-4" and "-6" options for forcing a specific address family instead of trying to autodetect. I have the feeling that the formats permitted by inet_aton() and inet_pton() are overlapping.

also this is a convention used by various ip related utilities (try ping -6)

This will be especially useful for stuff $(dns-resolve-name) where both address families could make sense.

Reply:
`inet_aton()` accepts:
 1. `a.b.c.d`
 2. `a.b.c`
 3. `a.b`
 4. `a`

_"components of the dotted address can be specified in decimal, octal (with a leading 0), or hexadecimal, with a leading 0X"_

whereas `inet_pton()` with `AF_INET6` accepts:
 1. `x:x:x:x:x:x:x:x`
 2. abbreviation of contigous 0s with: `::`
 3. `x:x:x:x:x:x:d.d.d.d`

---
So I do not really see an overlap, correct me if I am wrong.
If we can precisely support the auto-detection, then I do not see a definite reason to add `-4` and `-6` flags for this direction of the DNS resolvation. It would be nice to know the reason of the convention used by `ping` (and `ping6`) for example, I will check that!

Reply:
We concluded, that we will get back to these options, when `$(dns-resolve-name)` is implemented.